In [1]:
from collections import Counter
from imblearn.over_sampling import SMOTE
from sklearn import tree
from sklearn.model_selection import StratifiedShuffleSplit, StratifiedKFold
from sklearn import metrics
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [2]:
df = pd.read_csv("../../data/sea/dataset.csv")
df_1 = df.drop('title', axis=1)
df_1 = df_1.drop('abstract', axis=1)
df = df_1.drop('claim', axis=1)

In [3]:
x = df.loc[:, df.columns!='valid_patent'].astype(str)
y = df['valid_patent'].astype(int)
print(x.shape, y.shape)

(1905, 5566) (1905,)


In [4]:
x = np.array(x)
y = np.array(y)
def count_values(x):
    unique, counts = np.unique(x, return_counts=True)
    return dict(zip(unique, counts))

In [5]:
for i in range(1,100):
    
    total_accuracy = 0
    total_precision = 0
    total_recall = 0
    total_f1 = 0
    
    splits = i

    skf = StratifiedShuffleSplit(n_splits=splits, random_state=42, test_size=0.3)

    for train_index, test_index in skf.split(x, y):
    #     print("TRAIN:", train_index[:15], "\nTEST:", test_index[:15])    
        x_train, x_test = x[train_index], x[test_index]
        y_train, y_test = y[train_index], y[test_index]
    #     print("x_train: ", count_values(x_train), "x_test: ", count_values(x_test))
#         print("y_train: ", count_values(y_train), ", y_test: ", count_values(y_test))

#         print('\nOriginal dataset shape {}'.format(Counter(y_train)))
        sm = SMOTE(random_state=42)
        x_res , y_res = sm.fit_sample(x_train, y_train)
#         print('Resampled dataset shape {}'.format(Counter(y_res)))

        patent_tree = tree.DecisionTreeClassifier(criterion='gini', splitter='random', max_depth=5)
        patent_tree.fit(x_res, y_res)

        y_pred_tr = patent_tree.predict(x_test)

        accuracy = metrics.accuracy_score(y_test, y_pred_tr)
        total_accuracy += float(accuracy)

        precision = metrics.precision_score(y_test, y_pred_tr)
        total_precision += float(precision)

        recall = metrics.recall_score(y_test, y_pred_tr)
        total_recall += float(recall)

        f1_score = metrics.f1_score(y_test, y_pred_tr)
        total_f1 += float(f1_score)
            
#         if i % 10 == 0:
#             print('\nAccuracy: %.2f' % accuracy)
#             print('Precision: %.2f' % precision)
#             print('Recall: %.2f' % recall)
#             print('F1-score: %.2f' % f1_score)
#         #     print('Accuracy: %.2f' % metrics.accuracy_score(y_test, y_pred_tr))
#             print('==========================================================================\n')

    print('\n**'+str(i)+' splits Averge Score**')
    print('Accuracy: %.2f' % (total_accuracy/splits))
    print('Precision: %.2f' % (total_precision/splits))
    print('Recall: %.2f' % (total_recall/splits))
    print('F1-score: %.2f' % (total_f1/splits))
    print('*************************')


**1 splits Averge Score**
Accuracy: 0.97
Precision: 0.22
Recall: 0.44
F1-score: 0.30
*************************

**2 splits Averge Score**
Accuracy: 0.91
Precision: 0.07
Recall: 0.39
F1-score: 0.12
*************************

**3 splits Averge Score**
Accuracy: 0.96
Precision: 0.22
Recall: 0.22
F1-score: 0.18
*************************

**4 splits Averge Score**
Accuracy: 0.97
Precision: 0.38
Recall: 0.28
F1-score: 0.29
*************************

**5 splits Averge Score**
Accuracy: 0.97
Precision: 0.31
Recall: 0.40
F1-score: 0.33
*************************

**6 splits Averge Score**
Accuracy: 0.94
Precision: 0.31
Recall: 0.33
F1-score: 0.20
*************************

**7 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.43
F1-score: 0.23
*************************

**8 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.33
F1-score: 0.21
*************************

**9 splits Averge Score**
Accuracy: 0.96
Precision: 0.34
Recall: 0.41
F1-score: 0.28
******************

/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



**17 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.41
F1-score: 0.22
*************************

**18 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.48
F1-score: 0.24
*************************

**19 splits Averge Score**
Accuracy: 0.94
Precision: 0.14
Recall: 0.36
F1-score: 0.17
*************************

**20 splits Averge Score**
Accuracy: 0.95
Precision: 0.21
Recall: 0.35
F1-score: 0.20
*************************

**21 splits Averge Score**
Accuracy: 0.96
Precision: 0.23
Recall: 0.37
F1-score: 0.26
*************************

**22 splits Averge Score**
Accuracy: 0.94
Precision: 0.18
Recall: 0.42
F1-score: 0.23
*************************

**23 splits Averge Score**
Accuracy: 0.95
Precision: 0.25
Recall: 0.33
F1-score: 0.22
*************************

**24 splits Averge Score**
Accuracy: 0.94
Precision: 0.24
Recall: 0.44
F1-score: 0.24
*************************

**25 splits Averge Score**
Accuracy: 0.94
Precision: 0.15
Recall: 0.37
F1-score: 0.18
*********

/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


Recall: 0.39
F1-score: 0.21
*************************

**46 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.37
F1-score: 0.20
*************************

**47 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.43
F1-score: 0.21
*************************

**48 splits Averge Score**
Accuracy: 0.95
Precision: 0.26
Recall: 0.38
F1-score: 0.23
*************************

**49 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.39
F1-score: 0.23
*************************

**50 splits Averge Score**
Accuracy: 0.95
Precision: 0.22
Recall: 0.38
F1-score: 0.23
*************************

**51 splits Averge Score**
Accuracy: 0.94
Precision: 0.18
Recall: 0.42
F1-score: 0.21
*************************

**52 splits Averge Score**
Accuracy: 0.92
Precision: 0.16
Recall: 0.45
F1-score: 0.19
*************************


/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



**53 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.41
F1-score: 0.21
*************************

**54 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.39
F1-score: 0.20
*************************

**55 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.36
F1-score: 0.21
*************************

**56 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.39
F1-score: 0.23
*************************

**57 splits Averge Score**
Accuracy: 0.95
Precision: 0.24
Recall: 0.38
F1-score: 0.23
*************************

**58 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.45
F1-score: 0.24
*************************

**59 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.38
F1-score: 0.22
*************************

**60 splits Averge Score**
Accuracy: 0.95
Precision: 0.21
Recall: 0.36
F1-score: 0.21
*************************

**61 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.38

/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



F1-score: 0.21
*************************

**62 splits Averge Score**
Accuracy: 0.95
Precision: 0.23
Recall: 0.39
F1-score: 0.23
*************************

**63 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.40
F1-score: 0.22

/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



*************************

**64 splits Averge Score**
Accuracy: 0.93
Precision: 0.22
Recall: 0.41
F1-score: 0.21
*************************

**65 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.37
F1-score: 0.22
*************************

**66 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.42
F1-score: 0.22
*************************

**67 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.39
F1-score: 0.23
*************************

**68 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.42
F1-score: 0.23
*************************

**69 splits Averge Score**
Accuracy: 0.94
Precision: 0.21
Recall: 0.41
F1-score: 0.21
*************************

**70 splits Averge Score**
Accuracy: 0.94
Precision: 0.25
Recall: 0.42
F1-score: 0.25
*************************

**71 splits Averge Score**
Accuracy: 0.94
Precision: 0.19
Recall: 0.41
F1-score: 0.21
*************************

**72 splits Averge Score**
Accuracy: 0.94
Precision: 0.23
Recall: 0.

/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


F1-score: 0.23
*************************

**87 splits Averge Score**
Accuracy: 0.94
Precision: 0.18
Recall: 0.40
F1-score: 0.20
*************************

**88 splits Averge Score**
Accuracy: 0.95
Precision: 0.22
Recall: 0.38
F1-score: 0.23
*************************


/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



**89 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.40
F1-score: 0.22
*************************

**90 splits Averge Score**
Accuracy: 0.95
Precision: 0.21
Recall: 0.37
F1-score: 0.22
*************************

**91 splits Averge Score**
Accuracy: 0.93
Precision: 0.19
Recall: 0.45
F1-score: 0.21
*************************

**92 splits Averge Score**
Accuracy: 0.94
Precision: 0.20
Recall: 0.38
F1-score: 0.21
*************************


/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



**93 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.40
F1-score: 0.21
*************************

**94 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.42
F1-score: 0.23
*************************

**95 splits Averge Score**
Accuracy: 0.94
Precision: 0.22
Recall: 0.43
F1-score: 0.22
*************************

**96 splits Averge Score**
Accuracy: 0.93
Precision: 0.19
Recall: 0.45
F1-score: 0.22
*************************


/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/home/yehoon/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)



**97 splits Averge Score**
Accuracy: 0.94
Precision: 0.25
Recall: 0.41
F1-score: 0.24
*************************

**98 splits Averge Score**
Accuracy: 0.94
Precision: 0.21
Recall: 0.42
F1-score: 0.23
*************************

**99 splits Averge Score**
Accuracy: 0.94
Precision: 0.23
Recall: 0.41
F1-score: 0.23
*************************


| splits | accuracy | precision | recall | f1 score |
| :----: | :------: | :-------: | :----: | :------: |
|   10   |   0.94   |   0.18    |  0.41  |   0.22   |

y_train:  {0: 1313, 1: 20} , y_test:  {0: 563, 1: 9}

Original dataset shape Counter({0: 1313, 1: 20}) <br>
Resampled dataset shape Counter({0: 1313, 1: 1313})